ts/db_test.go

```diff
@@ -0,0 +1,351 @@
+		Read:  []string{"root"},
+		Write: []string{"root"},
+	}
+	testDefaultZoneConfig = proto.ZoneConfig{
+		ReplicaAttrs: []proto.Attributes{
+			{Attrs: []string{"dc1", "mem"}},
+			{Attrs: []string{"dc2", "mem"}},
+		},
+		RangeMinBytes: 1 << 10, // 1k
+		RangeMaxBytes: 1 << 18, // 256k
+		GC: &proto.GCPolicy{
+			TTLSeconds: 24 * 60 * 60, // 1 day
+		},
+	}
+)
+
+// tsTestModel is a model-based testing structure used to verify that time
+// series data sent to the Cockroach time series DB is stored correctly.
+//
+// This structure maintains a single ts.DB instance which stores data in a
+// monolithic Cockroach Store. It additionally maintains a simple in-memory key
+// value map, which is used as a model of the time series data stored in
+// Cockroach. The model maintains an expected copy of all keys beginning with
+// the time series data prefix.
+//
+// Each test should send a series of commands to the tsTestModel. Commands are
+// dispatched to the ts.DB instance, but are also used to modify the
+// in-memory key value model. Tests should periodically compare the in-memory
+// model to the actual data stored in the cockroach engine, ensuring that the
+// data matches.
+type tsTestModel struct {
+	t         testing.TB
+	modelData map[string]*proto.Value
+	db        *DB
+
+	client      *client.KV
+	engine      engine.Engine
+	store       *storage.Store
+	transport   multiraft.Transport
+	manualClock *hlc.ManualClock
+	clock       *hlc.Clock
+	stopper     *util.Stopper
+}
+
+// newTsTestModel creates a new tsTestModel instance. The Start() method must
+// be called before using it.
+func newTsTestModel(t *testing.T) *tsTestModel {
+	return &tsTestModel{
+		t:         t,
+		modelData: make(map[string]*proto.Value),
+	}
+}
+
+// Start constructs and starts the ts.DB instance and the cockroach store
+// backing it.
+func (tm *tsTestModel) Start() {
+	// Initialize some supporting objects needed to initialize the store.
+	rpcContext := rpc.NewContext(hlc.NewClock(hlc.UnixNano), rpc.LoadInsecureTLSConfig())
+	gossip := gossip.New(rpcContext, gossip.TestInterval, gossip.TestBootstrap)
+	tm.manualClock = hlc.NewManualClock(0)
+	tm.clock = hlc.NewClock(tm.manualClock.UnixNano)
+	tm.transport = multiraft.NewLocalRPCTransport()
+	tm.stopper = util.NewStopper()
+	tm.stopper.AddCloser(tm.transport)
+
+	// Initialize a new in-memory engine.
+	tm.engine = engine.NewInMem(proto.Attributes{Attrs: []string{"dc1", "mem"}}, 1<<20)
+
+	// Initialize and bootstrap a store which uses the engine.
+	sender := kv.NewLocalSender()
+	tm.store = storage.NewStore(tm.clock, tm.engine, client.NewKV(nil, sender), gossip, tm.transport, storage.TestStoreConfig)
+	if err := tm.store.Bootstrap(proto.StoreIdent{NodeID: 1, StoreID: 1}, tm.stopper); err != nil {
+		tm.t.Fatal(err)
+	}
+	if err := tm.store.BootstrapRange(); err != nil {
+		tm.t.Fatal(err)
+	}
+	sender.AddStore(tm.store)
+	if err := tm.store.Start(tm.stopper); err != nil {
+		tm.t.Fatal(err)
+	}
+
+	tm.initConfigs()
```

Comment:
I agree completely, to the point where I was obsessing over it in my first week.  I decided that I would come back to it after time series are in a better place.